source/language/classical.rst

```diff
@@ -39,12 +41,13 @@ assigned variable on the LHS.
 Classical bits and registers
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-Classical registers and bits support bitwise operators and the
-corresponding assignment operators between registers of the same size:
-and ``&``, or ``|``, xor ``^``. They support left shift ``<<`` and right shift ``>>`` by an unsigned
-integer, and the corresponding assignment operators. The shift operators
-shift bits off the end. They also support not ``~``, ``popcount`` [1]_, and left and
-right circular shift, ``rotl`` and ``rotr``, respectively.
+Classical registers, ``bit``, ``uint`` and ``angle`` support bitwise operators
+and the corresponding assignment operators with registers of the same size:
```

Comment:
I think that this is only true for ``uint``(and ``int``) and ``angle`` registers which have an explicit width attached. (This constraint is satisfied in the examples you provide.)

Reply:
The current spec doesn't allow `int` and `uint` without widths.  I know it's something we've discussed, but I guess it never got finished formally?  I don't actually remember if we'd fully agreed on what the allowed casts would be, and exactly what set of operations and function calls would be allowed, so perhaps that's the reason.

Reply:
You may be correct, in that we have not set down a formal specification of them: though early on in the TAC WG meetings, we seemed to consider the idea that they would exist, with enough seriousness to decide on a policy regarding bit-shifting of storage types with type `int` (without width specification).

I think the formal specification of these types, were pending a conclusion of what to do about implicit casting — we have considered various different things that these types could represent, and I think we may now have enough definition elsewhere in the type system to set those out. We may wish to flag this up separately as a topic to discuss, and return to this point once we've done so.

Reply:
When we added width-less bit-like types in openqasm/openqasm#276, we didn't update this text.  I can't remember what we'd discussed precisely in the types-and-casting working group, but looking back on it now, I'm not certain I see a need to forbid this within the current spec; we generally don't say anything about bounds checking, and that's where the issues might arise.

If we want to update this, I think it'd be in scope for a different PR - one focussed on how bounds-checking and out-of-bounds reads and writes are handled.